modules/database/src/ioc/rsrv/camessage.c

```diff
@@ -2434,24 +2438,77 @@ int camessage ( struct client *client )
+        /* disconnect clients that send an invalid command */
+        if (msg.m_cmmd > CA_PROTO_LAST_CMMD) {
+            log_header ( "CAS: Invalid command rejected",
+                client, &msg, 0, nmsg );
+            status = RSRV_ERROR;
+            break;
+        }
+
+        /*
+         * Calculate the message size, and update the pointer to the message payload area.
+         * Also, if the message is using the extended form, extract the the correct payload size
+         * and data count from the extended message header, and update m_postsize and m_count.
+         *
+         * Disconnect client that send an invalid payload size.
+         */
+        if (CA_V49(client->minor_version_number) && msg.m_postsize==0xffff && msg.m_count==0) {
             ca_uint32_t *pLW = ( ca_uint32_t * ) ( mp + 1 );
             if ( bytes_left < sizeof(*mp) + 2 * sizeof(*pLW) ) {
                 status = RSRV_OK;
                 break;
             }
-            msg.m_postsize  = ntohl ( pLW[0] );
-            msg.m_count     = ntohl ( pLW[1] );
+
+            /*
+             * For >= CA_V49 using the extended message form, the maximum message size is
+             * 0xffffffff bytes. The header is 0x18 bytes, so the maximum payload size is
+             * 0xffffffe7.
+             */
+            ca_uint32_t new_postsize = ntohl ( pLW[0] ); /* payload size on extended form headers */
+            if (new_postsize > 0xffffffe7) {
+                log_header ( "CAS: Invalid payload size rejected",
+                    client, &msg, 0, nmsg );
+                status = RSRV_ERROR;
+                break;
+            }
+
+            msg.m_postsize  = new_postsize;
+            msg.m_count     = ntohl ( pLW[1] ); /* Data count on extended form headers */
             msgsize = msg.m_postsize + sizeof(*mp) + 2 * sizeof ( *pLW );
             pBody = ( void * ) ( pLW + 2 );
         }
         else {
+            /*
+             * For < CA_V49, or higher but using the standard message form, the maximum
+             * message size is 0x4000 bytes. The header is 0x10 bytes, so the maximum
+             * payload size is 0x3ff0.
+             */
+            if (msg.m_postsize > 0x3ff0) {
+                log_header ( "CAS: Invalid payload size rejected",
+                    client, &msg, 0, nmsg );
+                status = RSRV_ERROR;
+                break;
+            }
             msgsize = msg.m_postsize + sizeof(*mp);
             pBody = ( void * ) ( mp + 1 );
         }
 
-        /* ignore deprecated clients, but let newer clients identify themselves. */
-        if (msg.m_cmmd!=CA_PROTO_VERSION && !CA_VSUPPORTED(client->minor_version_number)) {
+        /*
+         * Ignore deprecated clients, but let newer clients identify themselves.
+         *
+         * If "client->minor_version_number" is 0, it means that we haven't received the minor
+         * version number from the client yet, so we don't check the version.
```

Comment:
I think that is true only for >= `CA_V49` clients? Previous version of the protocol didn't contain the `CA_PROTO_VERSION` message, but instead the `CA_PROTO_NOOP` which doesn't include the minor version number in the header. Is that correct?

And `CA_VSUPPORTED()` will compare if the version is >= `CA_V44` instead.

Reply:
Well... does this mean that with 95fd255dd6b888b90624547165062b8a97278fab I actually broke support for clients before v4.9 (aka before 3.14)?

Reply:
Good question. I want to say no, but I'm not sure. 

If a  <`CA_V49`  client tries to send a command via TCP, it will be caught by the `CA_VSUPPORTED(client->minor_version_number)` statement as `client->minor_version_number = 0`. Versions previous to `CA_V49` (up to [R3.14.0-alpha2](https://github.com/epics-base/epics-base/releases/tag/R3.14.0-alpha2), which uses protocol minor [version 8](https://github.com/epics-base/epics-base/blob/R3.14.0-alpha2/src/ca/caProto.h#L39)) implements the [CA_PROTO_NOOP](https://github.com/epics-base/epics-base/blob/R3.14.0-alpha2/src/ca/caProto.h#L101) message instead; that message does not include the minor version number, but `0` instead, in the [header of the CA_PROTO_NOOP message](https://github.com/epics-base/epics-base/blob/R3.14.0-alpha2/src/ca/tcpiiu.cpp#L1007). [In fact, [the documentation says](https://github.com/epics-docs/epics-docs/blob/master/specs/ca_protocol.rst#ca_proto_version), that the version is not available for < `CA_V411`, but I think it should say `CA_V49` instead, see [epics-docs#2](https://github.com/epics-docs/epics-docs/issues/2)].

As the minor version number can be extracted from other types of messages (like for example the [CA_PROTO_CLAIM_CIU](https://github.com/epics-base/epics-base/blob/R3.14.0-alpha2/src/ca/tcpiiu.cpp#L1574), which correspond to the newer `CA_PROTO_CREATE_CHANNEL`, and the [CA_PROTO_SEARCH](https://github.com/epics-base/epics-base/blob/R3.14.0-alpha2/src/ca/nciu.cpp#L195) ), I added the extra condition `client->minor_version_number != 0` to allow those commands to be processed. 

This will however, let pass all commands until a version number is extracted; not  sure if that is better or worse.

Reply:
So... I was able to get a capture of a 3.13 era CA client searching,  The UDP packet has only CA_PROTO_SEARCH messages, so I think it would be ignored.  A fix, if we choose to make one, would be to expand this condition:

```c
if (msg.m_cmmd!=CA_PROTO_VERSION && !CA_VSUPPORTED(client->minor_version_number)) {
```

with exceptions for the other messages which include a the peer version.

Reply:
Yes. I think the original condition will ignore a `CA_PROTO_SEARCH` message, as `client->minor_version_number` will be `0`.  

I think your proposal of filtering by message type is a better solution respect to my original `client->minor_version_number != 0` condition, which will let pass all type of messages.

From the documentation, the messages that contain the minor version number, beside from `CA_PROTO_VERSION` of course, are: `CA_PROTO_SEARCH` and `CA_PROTO_CREATE_CHANNEL`. So, we could expand the condition to be:
```c
if (msg.m_cmmd!=CA_PROTO_VERSION &&
    msg.m_cmmd!=CA_PROTO_SEARCH &&
    msg.m_cmmd!=CA_PROTO_CREATE_CHANNEL &&
    !CA_VSUPPORTED(client->minor_version_number)) {
```

By the way, looking at the code, it seems like the version number is not being extracted from the `CA_PROTO_SEARCH` messages. It is extracted from a `CA_PROTO_CREATE_CHANNEL` message [here](https://github.com/epics-base/epics-base/blob/54e9d3f5d1256edf8e6eca06191f51ba0aaf4cf2/modules/database/src/ioc/rsrv/camessage.c#L1223), but the `CA_PROTO_SEARCH` messages only do a version check [here for TCP](https://github.com/epics-base/epics-base/blob/54e9d3f5d1256edf8e6eca06191f51ba0aaf4cf2/modules/database/src/ioc/rsrv/camessage.c#L2265-L2268) and [here for UDP](https://github.com/epics-base/epics-base/blob/54e9d3f5d1256edf8e6eca06191f51ba0aaf4cf2/modules/database/src/ioc/rsrv/camessage.c#L2178-L2181). So I think we need to add:
```c
client->minor_version_number = mp->m_count;
```
to extract version number from those messages as well.

Reply:
I'm not sure this would be enough.  In the >=3.14 CA traffic I've seen, the `CA_PROTO_CLIENT_NAME` and `CA_PROTO_HOST_NAME` message are sent before `CA_PROTO_CREATE_CHANNEL`.  The 3.13 capture I was able to get didn't include the TCP phase, so I don't know if these messages were present.

This situation was briefly discussed at the core meeting on the 24th.  The outcome was that, absent more developer interest, to recommend that sites still having 3.13 IOCs, and wishing to mix in base 7.0, to put the old IOCs behind a cagateway built against base 3.14 or 3.15.

So I don't think this PR needs to be so concerned with 3.13 compatibility as long as it isn't making the situation more complicated in case someone wants to revisit this topic in future.

Reply:
I reverted back this check to its original state.